--- openjdk/hotspot/src/share/vm/prims/jvm.cpp	2018-07-11 22:26:32.837214290 -0300
+++ openjdk/hotspot/src/share/vm/prims/jvm.cpp	2018-07-11 22:26:32.757213896 -0300
@@ -915,6 +915,8 @@
                                                h_prot, true, thread);
 
   if (result != NULL) {
+    from_class_oop = JNIHandles::resolve(from);
+    from_class = java_lang_Class::as_klassOop(from_class_oop);
     oop mirror = JNIHandles::resolve_non_null(result);
     klassOop to_class = java_lang_Class::as_klassOop(mirror);
     ClassLoaderDependencies::record_dependency(from_class, to_class, CHECK_NULL);
