Description: Allow zero to build on ppc64el
Author: Matthias Klose <doko@ubuntu.com>
Forwarded: no
Last-Update: 2014-07-06

Index: b/acinclude.m4
===================================================================
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -723,7 +723,7 @@ AC_DEFUN_ONCE([IT_ENABLE_ZERO_BUILD],
     arm|i386|ppc|s390|sh|sparc)
       ZERO_BITSPERWORD=32
       ;;
-    aarch64|alpha|amd64|ia64|ppc64|s390x|sparcv9)
+    aarch64|alpha|amd64|ia64|ppc64*|s390x|sparcv9)
       ZERO_BITSPERWORD=64
       ;;
     *)
