--- a/Makefile.am
+++ b/Makefile.am
@@ -2722,7 +2722,6 @@ check-langtools: stamps/jtreg.stamp
 	mkdir -p test/langtools/JTwork test/langtools/JTreport
 	$(BOOT_DIR)/bin/java -jar test/jtreg.jar -v1 -a -ignore:quiet \
 		-w:test/langtools/JTwork -r:test/langtools/JTreport \
-		$${ICEDTEA_JTREG_OTHERVM:--samevm} \
 		-jdk:$(BUILD_SDK_DIR) \
 		$(ICEDTEA_JTREG_OPTIONS) \
 		-exclude:$(abs_top_srcdir)/test/jtreg/excludelist.langtools.jtx \
@@ -2741,7 +2740,6 @@ endif
 	$${XVFB_RUN} \
 	$(BOOT_DIR)/bin/java -jar test/jtreg.jar -v1 -a -ignore:quiet \
 		-w:test/jdk/JTwork -r:test/jdk/JTreport \
-		$${ICEDTEA_JTREG_OTHERVM:--samevm} \
 		-jdk:$(BUILD_SDK_DIR) \
 		$(ICEDTEA_JTREG_OPTIONS) $(CACAO_EXCLUDE) \
 		-exclude:$(abs_top_srcdir)/test/jtreg/excludelist.jdk.jtx \
