--- openjdk/jdk/src/share/classes/sun/net/ftp/impl/FtpClient.java
+++ openjdk/jdk/src/share/classes/sun/net/ftp/impl/FtpClient.java
@@ -117,8 +117,8 @@ public class FtpClient extends sun.net.f
                 new PrivilegedAction<Object>() {
 
                     public Object run() {
-                        vals[0] = Integer.getInteger("sun.net.client.defaultReadTimeout", 0).intValue();
-                        vals[1] = Integer.getInteger("sun.net.client.defaultConnectTimeout", 0).intValue();
+                        vals[0] = Integer.getInteger("sun.net.client.defaultReadTimeout", 300_000).intValue();
+                        vals[1] = Integer.getInteger("sun.net.client.defaultConnectTimeout", 300_000).intValue();
                         encs[0] = System.getProperty("file.encoding", "ISO8859_1");
                         return null;
                     }
