--- openjdk-boot.orig/jdk/make/java/jar/Makefile	2015-04-10 15:45:22.924158352 +0100
+++ openjdk-boot/jdk/make/java/jar/Makefile	2015-07-13 21:00:08.802942914 +0100
@@ -26,6 +26,7 @@
 BUILDDIR = ../..
 PACKAGE = java.util.jar
 PRODUCT = sun
+OTHER_JAVACFLAGS += -Xprefer:source
 include $(BUILDDIR)/common/Defs.gmk
 
 #
--- openjdk-boot.orig/jdk/make/sun/security/other/Makefile	2015-04-10 15:45:23.332162376 +0100
+++ openjdk-boot/jdk/make/sun/security/other/Makefile	2015-07-13 21:43:46.637311660 +0100
@@ -26,6 +26,7 @@
 BUILDDIR = ../../..
 PACKAGE = sun.security.other
 PRODUCT = sun
+OTHER_JAVACFLAGS += -Xprefer:source
 include $(BUILDDIR)/common/Defs.gmk
 
 #
--- openjdk-boot.orig/jdk/make/sun/security/pkcs11/Makefile	2015-04-10 15:45:23.332162376 +0100
+++ openjdk-boot/jdk/make/sun/security/pkcs11/Makefile	2015-07-13 22:10:21.433512900 +0100
@@ -92,6 +92,7 @@
 PACKAGE = sun.security.pkcs11
 LIBRARY = j2pkcs11
 PRODUCT = sun
+OTHER_JAVACFLAGS += -Xprefer:source
 
 #
 # The following is for when we need to do postprocessing
--- openjdk-boot.orig/jdk/make/com/sun/crypto/provider/Makefile	2015-04-10 15:45:22.740156537 +0100
+++ openjdk-boot/jdk/make/com/sun/crypto/provider/Makefile	2015-07-13 22:28:28.630562057 +0100
@@ -95,6 +95,7 @@
 BUILDDIR = ../../../..
 PACKAGE = com.sun.crypto.provider
 PRODUCT = sun
+OTHER_JAVACFLAGS += -Xprefer:source
 
 #
 # The following is for when we need to do postprocessing
--- openjdk-boot.orig/jdk/make/sun/nio/cs/Makefile	2015-04-10 15:45:23.320162258 +0100
+++ openjdk-boot/jdk/make/sun/nio/cs/Makefile	2015-07-13 22:12:44.291022762 +0100
@@ -36,7 +36,7 @@
 # This re-directs all the class files to a separate location
 CLASSDESTDIR = $(TEMPDIR)/classes
 
-OTHER_JAVACFLAGS += -Xlint:serial,-deprecation
+OTHER_JAVACFLAGS += -Xlint:serial,-deprecation -Xprefer:source
 include $(BUILDDIR)/common/Defs.gmk
 
 #
